### ISSUE TITLE
Discussion on Data Model Extensions

### ISSUE THREAD

**Issue by vvolkl** [AIDASoft/podio#196]
# Discussion on Data Model Extensions

This has been brought up in the context of metadata  before -- but  there is only a blurry line between extensions of the data model and meta data (which I understand as information about the data itself -- examples would be bookkeeping information such as run numbers or readout strings from dd4hep that help interpret a cellid)
An example for extensions of the data model (mostly taking edm4hep as an example), on the other hand would be the data recorded by a hypothetical new hit detector, that, just to have a more concrete example, not only measures position and time, but also a real-valued quantity `Omega`. Another, less-fanciful example, are additional variables produced in reconstruction algorithms like b- and c- tags.

The software challenge for `Omega`- based reconstruction is that it should
* be possible to use the usual I/O facilities for `Omega` information too
* there may be parts of the `Omega` resonstruction code that only depend on the "plain" Hit and it should be able to use the plain hit for those.

In my experience users who want to write out an extra piece of information but can't will get very creative to so, so podio should have some answer to those example situations.
I see the following possible approaches.

## A: No extensions

In option A, the philosopy goes that anything that is really not covered by the datamodel should be added to the datamodel proper instead of just being an extension. The disadvantage is that especially with future detectors, it is very difficult to foresee all possible usecases, and that this could clutter up the data model and make it less clear to use overall.
In terms of implementation it is also not clear cut:

### A.1 Add completely new type

Adding a completely new type, say `TrackerHitWithOmega` that only shares the components (`Vector3f` ...) makes it difficult to refactor code in the Omega reconstruction.

```
  edm4hep::TrackerHitWithOmega:
    Description : "Tracker hit with Omega"
    Members :
      - unsigned long long cellID      //ID of the sensor that created this hit
      - int type                       //type of raw data hit, either one of edm4hep::TPCHIT, edm4hep::SIMTRACKERHIT - see collection parameters "TrackerHitTypeNames" and "TrackerHitTypeValues". 
      - int quality                    //quality bit flag of the hit.    
      - float time                     //time of the hit.
      - float eDep                     //energy deposited on the hit [GeV]. 
      - float eDepError                //error measured on EDep [GeV].
      - float edx                      //dE/dx of the hit in [GeV/mm]. 
      - edm4hep::Vector3d position     //hit position in [mm].
      - std::array<float,6> covMatrix  //covariance of the position (x,y,z), stored as lower triangle matrix. i.e. cov(x,x) , cov(y,x) , cov(y,y) , cov(z,x) , cov(z,y) , cov(z,z)
      - float omega     // futuristic new data
    VectorMembers:
      - edm4hep::ObjectID  rawHits     //raw data hits. Check getType to get actual data type. 
```

### A.2 Associate Omega with A TrackerHit

This implementation would have something like:

```
  edm4hep::TrackerHitWithOmegaAssoc:
    Description : "Tracker hit"
    Members :
      - float omega     // futuristic new data
    OneToManyRelations:
      - edm4hep::TrackerHit       hit     
```

and has some overhead due to the indices of the association. This was popular in fcc-edm: 
https://github.com/HEP-FCC/fcc-edm/blob/master/edm.yaml#L108

### A.3 Composition of Omega and TrackerHit

```
  edm4hep::TrackerHitWithOmegaComp:
    Description : "Tracker hit"
    Members :
      - float omega     // futuristic new data
      - edm4hep::TrackerHit hit
```
Besides that this does not work with podio yet as `TrackerHit` is not a component, this approach would have to implement some kind of inheritance in order for `TrackerHitWithOmegaCompCollection` to be used in interface requiring a `TrackerHitCollection`

## Option B: (Ab-)Use of Metadata

In principle, it's already possible to add a `vector<float>` of the same size as the collection proper to its collection metadata, but it's not very convenient to use (see k4SimDelphes / FCCAnalysis) and stretches the definition of Metadata.

## Option C: Unstructured Extensions

Podio could allow to write the Omega information as a `vector<float>` either through some extension interface or allowing direct access to the I/O. The connection to the TrackerHit is implicit through branchnames. This sacrifices some type safety, as Omega reconstruction would mostly work with `*(TrackerHitCollection, vector<float>)` interfaces.

## Option D: Data model extensions in other places

Same as A, but the new types are not directly added to the "original" .yaml, but possibly stores yaml snippet in the metadata and runs podio on the fly to generate the new classes.

Comments and other suggestions welcome!

**Comment by mirguest** [comment]
@vvolkl Thanks for the nice summary. 

About the option B, I am wondering that is that possible to introduce the "data type object" to describe the extended data model. 

In numpy, dtype [1] is used to interpret the bytes in the fixed size block. In reference [2], we can find the basic attributes:
* `char`: A unique character code for each of the 21 different built-in types
* `fields`: named fields
* `itemsize`: The element size of this data-type object

If the "data type object" is stored in the metadata part, then it is possible to use such information to interpret the extended data in the collections. 

[1] https://numpy.org/doc/stable/reference/arrays.dtypes.html
[2] https://numpy.org/doc/stable/reference/generated/numpy.dtype.html

**Comment by tmadlener** [comment]
Hi @vvolkl, thanks for bringing this up and for the summary.

Before starting I would like to add another question to the discussion: **"What should be possible with these extensions?"** I think that also has an influence on what can be a viable solution.

----
I would personally prefer something using some sort of composition like your A.3. It would agree nicely with the philosophy in podio in general, in my opinion. I also like the fact that it would more or less automatically give it a somewhat well defined interface that could at least enforce some invariants (e.g. there is always one `omega` per `TrackerHit` and the relation between the two).

Where I am not yet sure is whether we would need the possibility to have this extensions defined in yaml format, or if something similar could be achieved with templates. The problem I see with the yaml format is that you might need several essentially identical datatypes, simply because the additional data changes its name, e.g. another tracker hit where `omega` simply doesn't make any sense as the parameter name. Solving this by making the parameter name overly generic in the yaml to fit all use cases, makes this harder to use than it would need to be, in my opinion.

The following is somewhat imaginary syntax for a template solution, and I am not sure if something like this could actually be made to happen in c++, but it shows at least one of the problems that I encountered already now about the interface of accessing the additional value(s).

```cpp
// magic and unicorns make this work
using TrackerHitsWithFloat = podio::extend<TrackerHitCollection, float>;
 
// WRITING
// name of the float goes to the constructor here(?)
auto& trackerHitsWithOmega = store.create<TrackerHitsWithFloat>("amazingHits", "omega");

// use structured bindings to get access to everything(?)
auto& [trackerHit, omega] = trackerHitsWithOmega.create();

// READING
// Also here we need to know the name of the additional value
const auto& trackerHitsWithOmega = store.get<TrackerHitsWithFloat>("amazingHits", "omega");

// Looping could probably be done without "&" if the objects are small enough
for (const auto& [hit, omega] : trackerHitsWithOmega) { /**/ }
```

The main question is how to give access to the extension data in such an approach. The structured binding approach could work as long as there are not too many additional values, but it has its pitfalls (e.g. need to use `auto&` for being able to change `omega`). There could also be a `std::tuple` like index based approach, where probably an enum + utilities approach could make the indexing a bit easier. However, the nice `trackerHit.getOmega()` of the yaml version will not be possible, I think.

------
P.S. a small correction to Option B:

> In principle, it's already possible to add a `vector<float>` of the same size as the collection proper to its collection metadata

In the collection metadata one would need to store information about how to interpret the `vector<float>`, but the actual data would need to go into the "event metadata". This is because the collection metadata are assumed to not change after they have been read once (what happens on a file change is a different yet unsolved problem).

I think we should start to talk about "generic data" when we talk about extending the data model rather than "event metadata".

**Comment by tmadlener** [comment]
Is this still relevant now that we have the `podio::UserDataCollection`? Essentially it is now possible to store (more or less) arbitrary data along other collections, even though it is user responsibility to keep things synchronized, and the interface that was proposed above is not available.

**Comment by vvolkl** [comment]
I think this particular discussion is resolved with the `UserDataCollection`. Extending the .yaml might still have some usecases, as discussed in the key4hep meeting today, but we can shift that to a new issue.
